### PR TITLE
Config screen & Parser

### PR DESCRIPTION
**Pull request by xUser5000**
## What have been done
- Registered the plugin section in the settings screen.
- Added an icon (Graduation cap).
- Added .vscode directory to .gitignore
- Implemented a Parser for .bib files: I was surprised to find out that [Citation-js](https://github.com/larsgw/citation.js) does not have typescript definitions and I was not able to get around this issue, so I was faced with two choices: either to write the definition files myself or to abandon the library and look for something else. After exhaustive research, I found this open source [converter tool](https://bibtex.online/). It converts from BibTeX to APA. I looked into the source code and found this [file](https://github.com/enric1994/bibtexonline/blob/master/js/BibTex.js). This file does exactly the job of the parser class. In a nutshell, this library is a port of the [PEAR Structures_BibTex parser](http://pear.php.net/package/Structures_BibTex) written in PHP. In order to make porting the parser into javascript easier, the author has made use of many phpjs functions, which slightly puts a performance overhead but it is worth sacrificing and it can be improved later along the way. The library is meant to work in the browser, so I added a line at the end that just exports the main class: `module.exports = { BibTex };` just to make it compatible with other typescript modules. The library is used [here](https://github.com/enric1994/bibtexonline/blob/master/js/core.js) to parse the BibTeX data and also there are several useful functions to show APA in HTML format and convert it to other formats as well.

## Screenshots
Config Screen

### CODE REVIEW

File: src/ui/settings.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    	// Register the config screen page
+		await joplin.settings.registerSection("bibtex.settings", {
+			name: "bibtex",
+			label: "BibTeX Plugin",
+			description: "Use locally stored BibTeX files to include citations in Joplin notes",
+			iconName: "fa fa-graduation-cap"
+		});
+
+        // Bibtex file path
+		await joplin.settings.registerSetting('bibtex.file', {
```
Comment (laurent22) [inline src/ui/settings.ts:17]: This method is deprecated. Please use registerSettings.

**Comment by laurent22** [comment]
I don't quite get your reasoning for the bibtex lib. So which one is the best one? The one you had picked originally, or that weird PEAR PHP one? That what should drive the library you pick - it should be the best one. TypeScript definitions aren't an issue  - you can do without, it just means objects coming from that lib will be `any`, which is acceptable for output from third party libraries.

The problem of your approach is that it's a manual process to update the lib, not to mention that it's not clear by looking at the file where it comes from.

**Comment by xUser5000** [comment]
> TypeScript definitions aren't an issue - you can do without, it just means objects coming from that lib will be `any`, which is acceptable for output from third party libraries.

Yes, I thought so but for some reason, it just didn't work. I tried so many workarounds but still, the issue was not resolved. In this situation, I actually don't have a choice but to use this library.

**Comment by laurent22** [comment]
Please try again then and provide the error messages you're having. You've extracted some random JS file from 2008 that will never be updated just because you couldn't get a simple npm lib working. That's not good decision making here.

**Comment by xUser5000** [comment]
I did `npm install citation-js` and reverted the last commit on my local machine, then modified the following files:

`parser.util.ts`:

```js
import * as Cite from "citation-js";

export function parse (data: string) {
    let citation = new Cite(data, {
        format: 'string'
    });
    console.log(citation);
    return "Hello World";
}
```

`index.ts`:

```js
import joplin from 'api';
import { initConfigScreen } from './ui/settings';
import { parse } from './util/parser.util';

let dummyData = `
@article{Steinbeck2003,
	author = {Steinbeck, Christoph and Han, Yongquan and Kuhn, Stefan and Horlacher, Oliver and Luttmann, Edgar and Willighagen, Egon},
	year = {2003},
	title = {{The Chemistry Development Kit (CDK): an open-source Java library for Chemo- and Bioinformatics.}},
	journal = {Journal of chemical information and computer sciences},
	volume = {43},
	number = {2},
	pages = {493--500},
	doi = {10.1021/ci025584y},
	isbn = {2214707786},
	issn = {0095-2338},
	pmid = {12653513},
	url = {http://www.ncbi.nlm.nih.gov/pubmed/12653513}
  }
`;

joplin.plugins.register({
	onStart: async function() {
		console.info('BibTeX plugin started!');

		console.log(parse(dummyData));

		await initConfigScreen();
	},
});
```

Now, whenever I run `npm run dist`, I get the following log:
```bash
> joplin-plugin-bibtex@1.0.0 dist /home/xuser500/Projects/bibtex-plugin
> webpack --joplin-plugin-config buildMain && webpack --joplin-plugin-config buildExtraScripts && webpack --joplin-plugin-config createArchive

Child
    
    ERROR in ./node_modules/@citation-js/plugin-ris/lib-mjs/spec/mixed.js
    Module not found: Error: Can't resolve './additions' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-ris/lib-mjs/spec'
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/spec/mixed.js 2:0-36 3:26-35
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/spec/index.js
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/ris.js
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-bibtex/lib-mjs/mapping/biblatex.js
    Module not found: Error: Can't resolve './biblatexTypes' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-bibtex/lib-mjs/mapping'
     @ ./node_modules/@citation-js/plugin-bibtex/lib-mjs/mapping/biblatex.js 3:0-36 234:14-19 238:25-30
     @ ./node_modules/@citation-js/plugin-bibtex/lib-mjs/mapping/index.js
     @ ./node_modules/@citation-js/plugin-bibtex/lib-mjs/input/entries.js
     @ ./node_modules/@citation-js/plugin-bibtex/lib-mjs/input/index.js
     @ ./node_modules/@citation-js/plugin-bibtex/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-bibtex/lib/mapping/biblatex.js
    Module not found: Error: Can't resolve './biblatexTypes' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-bibtex/lib/mapping'
     @ ./node_modules/@citation-js/plugin-bibtex/lib/mapping/biblatex.js 12:44-70
     @ ./node_modules/@citation-js/plugin-bibtex/lib/mapping/index.js
     @ ./node_modules/@citation-js/plugin-bibtex/lib/input/entries.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-bibtex/lib-mjs/mapping/bibtex.js
    Module not found: Error: Can't resolve './bibtexTypes' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-bibtex/lib-mjs/mapping'
     @ ./node_modules/@citation-js/plugin-bibtex/lib-mjs/mapping/bibtex.js 3:0-34 195:19-24 209:19-24
     @ ./node_modules/@citation-js/plugin-bibtex/lib-mjs/mapping/index.js
     @ ./node_modules/@citation-js/plugin-bibtex/lib-mjs/input/entries.js
     @ ./node_modules/@citation-js/plugin-bibtex/lib-mjs/input/index.js
     @ ./node_modules/@citation-js/plugin-bibtex/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-bibtex/lib/mapping/bibtex.js
    Module not found: Error: Can't resolve './bibtexTypes' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-bibtex/lib/mapping'
     @ ./node_modules/@citation-js/plugin-bibtex/lib/mapping/bibtex.js 12:42-66
     @ ./node_modules/@citation-js/plugin-bibtex/lib/mapping/index.js
     @ ./node_modules/@citation-js/plugin-bibtex/lib/input/entries.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-ris/lib-mjs/index.js
    Module not found: Error: Can't resolve './config' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-ris/lib-mjs'
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/index.js 2:0-30 51:2-8
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-ris/lib-mjs/ris.js
    Module not found: Error: Can't resolve './config' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-ris/lib-mjs'
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/ris.js 2:0-30 104:29-35
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-wikidata/lib-mjs/index.js
    Module not found: Error: Can't resolve './config' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-wikidata/lib-mjs'
     @ ./node_modules/@citation-js/plugin-wikidata/lib-mjs/index.js 7:0-30 103:2-8
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-wikidata/lib-mjs/id.js
    Module not found: Error: Can't resolve './config' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-wikidata/lib-mjs'
     @ ./node_modules/@citation-js/plugin-wikidata/lib-mjs/id.js 2:0-30 6:54-60
     @ ./node_modules/@citation-js/plugin-wikidata/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-wikidata/lib-mjs/prop.js
    Module not found: Error: Can't resolve './config' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-wikidata/lib-mjs'
     @ ./node_modules/@citation-js/plugin-wikidata/lib-mjs/prop.js 4:0-30 122:15-21
     @ ./node_modules/@citation-js/plugin-wikidata/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-ris/lib-mjs/ris.js
    Module not found: Error: Can't resolve './dataTypes' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-ris/lib-mjs'
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/ris.js 5:0-37 13:28-38 14:37-47
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-bibtex/lib/config.js
    Module not found: Error: Can't resolve './mapping/biblatexTypes' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-bibtex/lib'
     @ ./node_modules/@citation-js/plugin-bibtex/lib/config.js 8:39-73
     @ ./node_modules/@citation-js/plugin-bibtex/lib/input/entries.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-bibtex/lib-mjs/config.js
    Module not found: Error: Can't resolve './mapping/biblatexTypes' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-bibtex/lib-mjs'
     @ ./node_modules/@citation-js/plugin-bibtex/lib-mjs/config.js 1:0-52 7:4-12
     @ ./node_modules/@citation-js/plugin-bibtex/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-bibtex/lib/config.js
    Module not found: Error: Can't resolve './mapping/bibtexTypes' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-bibtex/lib'
     @ ./node_modules/@citation-js/plugin-bibtex/lib/config.js 10:37-69
     @ ./node_modules/@citation-js/plugin-bibtex/lib/input/entries.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-bibtex/lib-mjs/config.js
    Module not found: Error: Can't resolve './mapping/bibtexTypes' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-bibtex/lib-mjs'
     @ ./node_modules/@citation-js/plugin-bibtex/lib-mjs/config.js 2:0-48 8:4-10
     @ ./node_modules/@citation-js/plugin-bibtex/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-ris/lib-mjs/spec/index.js
    Module not found: Error: Can't resolve './new' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-ris/lib-mjs/spec'
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/spec/index.js 2:0-25 6:7-11
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/ris.js
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-ris/lib-mjs/spec/mixed.js
    Module not found: Error: Can't resolve './new' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-ris/lib-mjs/spec'
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/spec/mixed.js 1:0-24 3:15-18
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/spec/index.js
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/ris.js
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-wikidata/lib-mjs/entity.js
    Module not found: Error: Can't resolve './props' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-wikidata/lib-mjs'
     @ ./node_modules/@citation-js/plugin-wikidata/lib-mjs/entity.js 4:0-46 45:21-26 46:31-36 58:16-28
     @ ./node_modules/@citation-js/plugin-wikidata/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-ris/lib-mjs/converters.js
    Module not found: Error: Can't resolve './spec/types' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-ris/lib-mjs'
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/converters.js 2:0-33 124:13-18 128:13-18
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/ris.js
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-ris/lib-mjs/spec/index.js
    Module not found: Error: Can't resolve './types' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-ris/lib-mjs/spec'
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/spec/index.js 1:0-28 10:0-17
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/ris.js
     @ ./node_modules/@citation-js/plugin-ris/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/@citation-js/plugin-wikidata/lib-mjs/prop.js
    Module not found: Error: Can't resolve './types' in '/home/xuser500/Projects/bibtex-plugin/node_modules/@citation-js/plugin-wikidata/lib-mjs'
     @ ./node_modules/@citation-js/plugin-wikidata/lib-mjs/prop.js 5:0-28 110:7-12 115:9-14
     @ ./node_modules/@citation-js/plugin-wikidata/lib-mjs/index.js
     @ ./node_modules/citation-js/index.js
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
    
    ERROR in ./node_modules/citation-js/index.js
    Module not found: Error: Can't resolve '@citation-js/plugin-bibtex/lib/mapping/bibtexTypes' in '/home/xuser500/Projects/bibtex-plugin/node_modules/citation-js'
     @ ./node_modules/citation-js/index.js 144:4-65 195:4-65
     @ ./src/util/parser.util.ts
     @ ./src/index.ts
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! joplin-plugin-bibtex@1.0.0 dist: `webpack --joplin-plugin-config buildMain && webpack --joplin-plugin-config buildExtraScripts && webpack --joplin-plugin-config createArchive`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the joplin-plugin-bibtex@1.0.0 dist script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/xuser500/.npm/_logs/2021-06-07T16_02_12_332Z-debug.log
```

**Comment by xUser5000** [comment]
I tried most if not all workarounds possible. But still :"

**Comment by xUser5000** [comment]
One other thing, when I try to call `registerSettings()` instead of the deprecated one, I get the following:
`Property 'registerSettings' does not exist on type 'JoplinSettings'. Did you mean 'registerSetting'?`
Indeed I the API documentation recommends using the new method, it does not seem to be in the declaration file. Should I try to manually update the `JoplinSettings.d.ts` file ?

**Comment by laurent22** [comment]
Ok it looks like it'd take extra efforts to get it working (there's a thread about it in their issue tracker) but I don't know if it's worth it. So you might be right and should use whatever works for now since it looks like all bibtex modules are pretty bad anyway. However make sure your code is not too tightly coupled to the library so that you can potentially change it if needed.

**Comment by xUser5000** [comment]
> One other thing, when I try to call `registerSettings()` instead of the deprecated one, I get the following:
> `Property 'registerSettings' does not exist on type 'JoplinSettings'. Did you mean 'registerSetting'?`
> Indeed, the API documentation recommends using the new method. However, it does not seem to be in the declaration file. Should I try to manually update the `JoplinSettings.d.ts` file ?

What about this one?

**Comment by xUser5000** [comment]
What about now?

**Comment by laurent22** [comment]
Looks good now, let's merge.

Once you start playing with parsing BibTex files I'd suggest you also setup test units, so that you can check that the lib properly handles a few cases you want to support.

**Pull request merged**